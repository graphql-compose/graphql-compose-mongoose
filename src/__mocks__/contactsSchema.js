/* @flow */

import { Schema } from './mongooseCommon';

const ContactsSchema = new Schema({
  phones: [String],
  email: String,
  skype: String,
  locationId: Schema.Types.ObjectId,
});

export default ContactsSchema;
