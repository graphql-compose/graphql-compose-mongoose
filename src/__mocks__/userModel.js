/* @flow */

import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'UserModel',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'UserModel',
    },

    name: {
      type: String,
      description: 'Person name',
    },

    age: {
      type: Number,
      description: 'Full years',
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    // createdAt, created via option `timastamp: true` (see bottom)
    // updatedAt, created via option `timastamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

UserSchema.virtual('nameVirtual').get(function () { // eslint-disable-line
  return `VirtualFieldValue${this._id}`;
});

const UserModel = mongoose.model('User', UserSchema);

export { UserSchema, UserModel };
