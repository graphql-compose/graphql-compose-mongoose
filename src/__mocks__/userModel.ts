import { mongoose, Schema } from './mongooseCommon';
import ContactsSchema from './contactsSchema';
import enumEmployment from './enumEmployment';
import LanguageSchema from './languageSchema';
import { Document } from 'mongoose';

const UserSchema = new Schema(
  {
    subDoc: {
      field1: String,
      field2: {
        field21: String,
        field22: String,
      },
    },

    user: {
      type: Schema.Types.ObjectId,
      ref: 'User',
    },

    users: {
      type: [Schema.Types.ObjectId],
      ref: 'User',
    },

    n: {
      type: String,
      required: true,
      description: 'Person name',
      alias: 'name',
    },

    age: {
      type: Number,
      description: 'Full years',
      required() {
        // in graphql this field should be Nullable
        return (this as any).name === 'Something special';
      },
    },

    gender: {
      type: String,
      enum: ['male', 'female', 'ladyboy'],
    },

    skills: {
      type: [String],
      default: [],
      description: 'List of skills',
    },

    employment: {
      type: [
        {
          type: String,
          enum: Object.keys(enumEmployment),
        },
      ],
      description: 'List of desired employment types',
      index: true,
    },

    relocation: {
      type: Boolean,
      description: 'Does candidate relocate to another region',
    },

    contacts: {
      type: ContactsSchema,
      default: {},
      description: 'Contacts of person (phone, skype, mail and etc)',
    },

    languages: {
      type: [LanguageSchema],
      default: [],
      description: 'Knowledge of languages',
    },

    __secretField: {
      type: String,
    },

    someDynamic: {
      type: Schema.Types.Mixed,
      description: "Some mixed value, that served with @taion's `graphql-type-json`",
    },

    periods: [{ from: Number, to: Number }],

    someDeep: {
      periods: [{ from: Number, to: Number }],
    },

    salary: {
      type: Schema.Types.Decimal128,
    },

    mapField: {
      type: Map,
      of: String,
    },

    mapFieldDeep: {
      subField: {
        type: Map,
        of: String,
      },
    },

    billingAddress: {
      street: { type: String, index: true },
      state: { type: String, index: true, enum: ['FL', 'MA', 'NY'] },
      country: { type: String, index: true },
    },

    // for error payloads tests
    valid: {
      type: String,
      required: false,
      validate: [
        () => {
          return false;
        },
        'this is a validate message',
      ],
    },

    // createdAt, created via option `timestamp: true` (see bottom)
    // updatedAt, created via option `timestamp: true` (see bottom)
  },
  {
    timestamps: true, // add createdAt, updatedAt fields
    toJSON: { getters: true },
    toObject: { virtuals: true },
  }
);

UserSchema.set('autoIndex', false);
UserSchema.index({ name: 1, age: -1 });

// eslint-disable-next-line
UserSchema.virtual('nameVirtual').get(function (this: any) {
  return `VirtualFieldValue${this._id}`;
});

export interface IUser extends Document {
  _id: any;
  name?: string;
  age?: number;
  gender?: string;
  someDynamic?: any;
  skills?: string[];
  relocation?: boolean;
  contacts: {
    email: string;
    skype?: string;
  };
}

const UserModel = mongoose.model<IUser>('User', UserSchema);

export { UserSchema, UserModel };
