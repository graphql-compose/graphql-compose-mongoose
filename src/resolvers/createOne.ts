import { Resolver, ObjectTypeComposer } from 'graphql-compose';
import type { Model, Document } from 'mongoose';
import { recordHelperArgs, RecordHelperArgsOpts } from './helpers';
import type { ExtendedResolveParams } from './index';
import { addErrorCatcherField } from './helpers/errorCatcher';
import { validateAndThrow } from './helpers/validate';
import { PayloadRecordIdHelperOpts, payloadRecordId } from './helpers/payloadRecordId';

export interface CreateOneResolverOpts {
  /** If you want to generate different resolvers you may avoid Type name collision by adding a suffix to type names */
  suffix?: string;
  /** Customize input-type for `record` argument */
  record?: RecordHelperArgsOpts;
  /** Customize payload.recordId field. If false, then this field will be removed. */
  recordId?: PayloadRecordIdHelperOpts | false;
}

type TArgs = {
  record: Record<string, any>;
};

export function createOne<TSource = any, TContext = any, TDoc extends Document = any>(
  model: Model<TDoc>,
  tc: ObjectTypeComposer<TDoc, TContext>,
  opts?: CreateOneResolverOpts
): Resolver<TSource, TContext, TArgs, TDoc> {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver createOne() should be instance of Mongoose Model.');
  }

  if (!tc || tc.constructor.name !== 'ObjectTypeComposer') {
    throw new Error(
      'Second arg for Resolver createOne() should be instance of ObjectTypeComposer.'
    );
  }

  const tree = model.schema.obj;
  const requiredFields = [];
  for (const field in tree) {
    if (tree.hasOwnProperty(field)) {
      const fieldOptions = tree[field];
      if (fieldOptions.required && typeof fieldOptions.required !== 'function') {
        requiredFields.push(field);
      }
    }
  }

  const outputTypeName = `CreateOne${tc.getTypeName()}${opts?.suffix || ''}Payload`;
  const outputType = tc.schemaComposer.getOrCreateOTC(outputTypeName, (t) => {
    t.setFields({
      ...payloadRecordId(tc, opts?.recordId),
      record: {
        type: tc,
        description: 'Created document',
      },
    });
  });

  const resolver = tc.schemaComposer.createResolver<TSource, TArgs>({
    name: `createOne${opts?.suffix || ''}`,
    kind: 'mutation',
    description: 'Create one document with mongoose defaults, setters, hooks and validation',
    type: outputType,
    args: {
      ...recordHelperArgs(tc, {
        prefix: 'CreateOne',
        suffix: `${opts?.suffix || ''}Input`,
        removeFields: ['id', '_id'],
        isRequired: true,
        requiredFields,
        ...opts?.record,
      }),
    },
    resolve: (async (resolveParams: ExtendedResolveParams<TDoc>) => {
      const recordData = resolveParams?.args?.record;

      if (!(typeof recordData === 'object') || Object.keys(recordData).length === 0) {
        throw new Error(
          `${tc.getTypeName()}.createOne resolver requires at least one value in args.record`
        );
      }

      let doc = new model(recordData);
      if (resolveParams.beforeRecordMutate) {
        doc = await resolveParams.beforeRecordMutate(doc, resolveParams);
        if (!doc) return null;
      }

      await validateAndThrow(doc);
      await doc.save({ validateBeforeSave: false });

      return {
        record: doc,
      };
    }) as any,
  });

  // Add `error` field to payload which can catch resolver Error
  // and return it in mutation payload
  addErrorCatcherField(resolver);

  return resolver;
}
