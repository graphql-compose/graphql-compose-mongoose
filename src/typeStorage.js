/* @flow */

import { isFunction } from 'graphql-compose';

class TypeStorage extends Map {
  // this `create` hack due TypeError:
  // Constructor Map requires 'new' at TypeStorage.Map (native)
  static create(array?: any[]): TypeStorage {
    const inst = new Map(array);
    // $FlowFixMe
    inst.__proto__ = TypeStorage.prototype; // eslint-disable-line
    // $FlowFixMe
    return inst;
  }

  getOrSet<T>(typeName: string, typeOrThunk: T | (() => T)): T {
    if (this.has(typeName)) {
      // $FlowFixMe
      return this.get(typeName);
    }

    // $FlowFixMe
    const gqType: T = isFunction(typeOrThunk) ? typeOrThunk() : typeOrThunk;
    if (gqType) {
      this.set(typeName, gqType);
    }

    return gqType;
  }
}

const typeStorage = TypeStorage.create();
export default typeStorage;
