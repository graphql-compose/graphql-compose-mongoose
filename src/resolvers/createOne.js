/* @flow */
/* eslint-disable no-param-reassign, new-cap */

import { Resolver, TypeComposer } from 'graphql-compose';
import { recordHelperArgs } from './helpers/record';
import typeStorage from '../typeStorage';
import GraphQLMongoID from '../types/mongoid';
import type { MongooseModelT, ExtendedResolveParams, GenResolverOpts } from '../definition';

export default function createOne(
  model: MongooseModelT,
  typeComposer: TypeComposer,
  opts?: GenResolverOpts
): Resolver<*, *> {
  if (!model || !model.modelName || !model.schema) {
    throw new Error('First arg for Resolver createOne() should be instance of Mongoose Model.');
  }

  if (!(typeComposer instanceof TypeComposer)) {
    throw new Error('Second arg for Resolver createOne() should be instance of TypeComposer.');
  }

  const outputTypeName = `CreateOne${typeComposer.getTypeName()}Payload`;
  const outputType = typeStorage.getOrSet(
    outputTypeName,
    TypeComposer.create({
      name: outputTypeName,
      fields: {
        recordId: {
          type: GraphQLMongoID,
          description: 'Created document ID',
        },
        record: {
          type: typeComposer.getType(),
          description: 'Created document',
        },
      },
    })
  );

  const resolver = new Resolver({
    name: 'createOne',
    kind: 'mutation',
    description: 'Create one document with mongoose defaults, setters, hooks and validation',
    type: outputType,
    args: {
      ...recordHelperArgs(typeComposer, {
        recordTypeName: `CreateOne${typeComposer.getTypeName()}Input`,
        removeFields: ['id', '_id'],
        isRequired: true,
        ...(opts && opts.record),
      }),
    },
    resolve: (resolveParams: ExtendedResolveParams) => {
      const recordData = (resolveParams.args && resolveParams.args.record) || {};

      if (!(typeof recordData === 'object') || Object.keys(recordData).length === 0) {
        return Promise.reject(
          new Error(
            `${typeComposer.getTypeName()}.createOne resolver requires ` +
              'at least one value in args.record'
          )
        );
      }

      // $FlowFixMe
      return Promise.resolve(new model(recordData))
        .then(doc => {
          if (resolveParams.beforeRecordMutate) {
            return resolveParams.beforeRecordMutate(doc, resolveParams);
          }
          return doc;
        })
        .then(doc => doc.save())
        .then(record => {
          if (record) {
            return {
              record,
              recordId: typeComposer.getRecordIdFn()(record),
            };
          }

          return null;
        });
    },
  });

  return resolver;
}
