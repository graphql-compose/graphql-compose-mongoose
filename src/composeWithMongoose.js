/* @flow */
/* eslint-disable no-use-before-define, no-param-reassign, global-require */

import { TypeComposer, InputTypeComposer } from 'graphql-compose';
import { convertModelToGraphQL } from './fieldsConverter';
import * as resolvers from './resolvers';
import { getUniqueIndexes, extendByReversedIndexes } from './utils/getIndexesFromModel';

import type {
  MongooseModelT,
  TypeConverterOpts,
  TypeConverterResolversOpts,
  TypeConverterInputTypeOpts,
  ConnectionSortMapOpts,
  PaginationOpts,
} from './definition';

export function composeWithMongoose(
  model: Object, // MongooseModelT, TODO use Model from mongoose_v4.x.x definition when it will be public
  opts: TypeConverterOpts = {}
): TypeComposer {
  const name: string = (opts && opts.name) || model.modelName;

  const typeComposer = convertModelToGraphQL(model, name);

  if (opts.description) {
    typeComposer.setDescription(opts.description);
  }

  if (opts.fields) {
    prepareFields(typeComposer, opts.fields);
  }

  typeComposer.setRecordIdFn(source => (source ? `${source._id}` : ''));

  createInputType(typeComposer, opts.inputType);

  if (!{}.hasOwnProperty.call(opts, 'resolvers') || opts.resolvers !== false) {
    createResolvers(model, typeComposer, opts.resolvers || {});
  }

  return typeComposer;
}

export function prepareFields(
  typeComposer: TypeComposer,
  opts: {
    only?: string[],
    remove?: string[],
  }
) {
  if (Array.isArray(opts.only)) {
    const onlyFieldNames: string[] = opts.only;
    const removeFields = Object.keys(typeComposer.getFields()).filter(
      fName => onlyFieldNames.indexOf(fName) === -1
    );
    typeComposer.removeField(removeFields);
  }
  if (opts.remove) {
    typeComposer.removeField(opts.remove);
  }
}

export function prepareInputFields(
  inputTypeComposer: InputTypeComposer,
  inputFieldsOpts: {
    only?: string[],
    remove?: string[],
    required?: string[],
  }
) {
  if (Array.isArray(inputFieldsOpts.only)) {
    const onlyFieldNames: string[] = inputFieldsOpts.only;
    const removeFields = Object.keys(inputTypeComposer.getFields()).filter(
      fName => onlyFieldNames.indexOf(fName) === -1
    );
    inputTypeComposer.removeField(removeFields);
  }
  if (inputFieldsOpts.remove) {
    inputTypeComposer.removeField(inputFieldsOpts.remove);
  }
  if (inputFieldsOpts.required) {
    inputTypeComposer.makeRequired(inputFieldsOpts.required);
  }
}

export function createInputType(
  typeComposer: TypeComposer,
  inputTypeOpts?: TypeConverterInputTypeOpts = {}
): void {
  const inputTypeComposer = typeComposer.getInputTypeComposer();

  if (inputTypeOpts.name) {
    inputTypeComposer.setTypeName(inputTypeOpts.name);
  }

  if (inputTypeOpts.description) {
    inputTypeComposer.setDescription(inputTypeOpts.description);
  }

  if (inputTypeOpts.fields) {
    prepareInputFields(inputTypeComposer, inputTypeOpts.fields);
  }
}

export function createResolvers(
  model: MongooseModelT,
  typeComposer: TypeComposer,
  opts: TypeConverterResolversOpts
): void {
  const names = resolvers.getAvailableNames();
  names.forEach(resolverName => {
    if (!{}.hasOwnProperty.call(opts, resolverName) || opts[resolverName] !== false) {
      const createResolverFn = resolvers[resolverName];
      if (createResolverFn) {
        const resolver = createResolverFn(model, typeComposer, opts[resolverName] || {});
        typeComposer.setResolver(resolverName, resolver);
      }
    }
  });

  if (!{}.hasOwnProperty.call(opts, 'connection') || opts.connection !== false) {
    prepareConnectionResolver(model, typeComposer, opts.connection ? opts.connection : {});
  }

  if (!{}.hasOwnProperty.call(opts, 'pagination') || opts.pagination !== false) {
    preparePaginationResolver(typeComposer, opts.pagination || {});
  }
}

export function preparePaginationResolver(typeComposer: TypeComposer, opts: PaginationOpts) {
  try {
    require.resolve('graphql-compose-pagination');
  } catch (e) {
    return;
  }
  const composeWithPagination = require('graphql-compose-pagination').default;
  composeWithPagination(typeComposer, {
    findResolverName: 'findMany',
    countResolverName: 'count',
    ...opts,
  });
}

export function prepareConnectionResolver(
  model: MongooseModelT,
  typeComposer: TypeComposer,
  opts: ConnectionSortMapOpts
) {
  try {
    require.resolve('graphql-compose-connection');
  } catch (e) {
    return;
  }
  const composeWithConnection = require('graphql-compose-connection').default;

  const uniqueIndexes = extendByReversedIndexes(getUniqueIndexes(model), {
    reversedFirst: true,
  });
  const sortConfigs = {};
  uniqueIndexes.forEach(indexData => {
    const keys = Object.keys(indexData);
    let name = keys.join('__').toUpperCase().replace(/[^_a-zA-Z0-9]/i, '__');
    if (indexData[keys[0]] === 1) {
      name = `${name}_ASC`;
    } else if (indexData[keys[0]] === -1) {
      name = `${name}_DESC`;
    }
    sortConfigs[name] = {
      value: indexData,
      cursorFields: keys,
      beforeCursorQuery: (rawQuery, cursorData) => {
        keys.forEach(k => {
          if (!rawQuery[k]) rawQuery[k] = {};
          if (indexData[k] === 1) {
            rawQuery[k].$lt = cursorData[k];
          } else {
            rawQuery[k].$gt = cursorData[k];
          }
        });
      },
      afterCursorQuery: (rawQuery, cursorData) => {
        keys.forEach(k => {
          if (!rawQuery[k]) rawQuery[k] = {};
          if (indexData[k] === 1) {
            rawQuery[k].$gt = cursorData[k];
          } else {
            rawQuery[k].$lt = cursorData[k];
          }
        });
      },
    };
  });

  composeWithConnection(typeComposer, {
    findResolverName: 'findMany',
    countResolverName: 'count',
    sort: {
      ...sortConfigs,
      ...opts,
    },
  });
}
