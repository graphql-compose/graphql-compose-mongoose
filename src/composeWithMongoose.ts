/* eslint-disable no-use-before-define, no-param-reassign, global-require */

import type { ObjectTypeComposer, InputTypeComposer, SchemaComposer } from 'graphql-compose';
import { schemaComposer as globalSchemaComposer } from 'graphql-compose';
import type { Model, Document } from 'mongoose';
import { convertModelToGraphQL } from './fieldsConverter';
import { allResolvers, AllResolversOpts } from './resolvers';
import MongoID from './types/MongoID';
import { GraphQLResolveInfo } from 'graphql';

export type ComposeWithMongooseOpts<TContext> = {
  schemaComposer?: SchemaComposer<TContext>;
  name?: string;
  description?: string;
  fields?: {
    only?: string[];
    // rename?: { [oldName: string]: string },
    remove?: string[];
  };
  inputType?: TypeConverterInputTypeOpts;
  resolvers?: false | AllResolversOpts;
  /** You may customize document id */
  transformRecordId?: TransformRecordIdFn<TContext>;
};

export type TransformRecordIdFn<TContext = any> = (
  source: Document,
  context: TContext,
  info: GraphQLResolveInfo
) => any;

export type TypeConverterInputTypeOpts = {
  name?: string;
  description?: string;
  fields?: {
    only?: string[];
    remove?: string[];
    required?: string[];
  };
};

export function composeWithMongoose<TDoc extends Document, TContext = any>(
  model: Model<TDoc>,
  opts: ComposeWithMongooseOpts<TContext> = {}
): ObjectTypeComposer<TDoc, TContext> {
  const m: Model<any> = model;
  const name: string = (opts && opts.name) || m.modelName;

  const sc = opts.schemaComposer || globalSchemaComposer;
  sc.add(MongoID);

  if (sc.has(name)) {
    throw new Error(
      `You try to generate GraphQL Type with name ${name} from mongoose model but this type already exists in SchemaComposer. Please choose another type name "composeWithMongoose(model, { name: 'NewTypeName' })", or reuse existed type "schemaComposer.getOTC('TypeName')", or remove type from SchemaComposer before calling composeWithMongoose method "schemaComposer.delete('TypeName')".`
    );
  }
  if (sc.has(m.schema)) {
    // looks like you want to generate new TypeComposer from model
    // so remove cached model (which is used for cross-reference types)
    sc.delete(m.schema);
  }

  const tc = convertModelToGraphQL(m, name, sc);

  if (opts.description) {
    tc.setDescription(opts.description);
  }

  if (opts.fields) {
    prepareFields(tc, opts.fields);
  }

  createInputType(tc, opts.inputType);

  if (!{}.hasOwnProperty.call(opts, 'resolvers') || opts.resolvers !== false) {
    createResolvers(m, tc, opts.resolvers || {});
  }

  tc.makeFieldNonNull('_id');

  return tc;
}

export function prepareFields(
  tc: ObjectTypeComposer<any, any>,
  opts: {
    only?: string[];
    remove?: string[];
  }
): void {
  if (Array.isArray(opts.only)) {
    const onlyFieldNames: string[] = opts.only;
    const removeFields = Object.keys(tc.getFields()).filter(
      (fName) => onlyFieldNames.indexOf(fName) === -1
    );
    tc.removeField(removeFields);
  }
  if (opts.remove) {
    tc.removeField(opts.remove);
  }
}

export function prepareInputFields(
  inputTypeComposer: InputTypeComposer<any>,
  inputFieldsOpts: {
    only?: string[];
    remove?: string[];
    required?: string[];
  }
): void {
  if (Array.isArray(inputFieldsOpts.only)) {
    const onlyFieldNames: string[] = inputFieldsOpts.only;
    const removeFields = Object.keys(inputTypeComposer.getFields()).filter(
      (fName) => onlyFieldNames.indexOf(fName) === -1
    );
    inputTypeComposer.removeField(removeFields);
  }
  if (inputFieldsOpts.remove) {
    inputTypeComposer.removeField(inputFieldsOpts.remove);
  }
  if (inputFieldsOpts.required) {
    inputTypeComposer.makeFieldNonNull(inputFieldsOpts.required);
  }
}

export function createInputType(
  tc: ObjectTypeComposer<any, any>,
  inputTypeOpts: TypeConverterInputTypeOpts = {}
): void {
  const inputTypeComposer = tc.getInputTypeComposer();

  if (inputTypeOpts.name) {
    inputTypeComposer.setTypeName(inputTypeOpts.name);
  }

  if (inputTypeOpts.description) {
    inputTypeComposer.setDescription(inputTypeOpts.description);
  }

  if (inputTypeOpts.fields) {
    prepareInputFields(inputTypeComposer, inputTypeOpts.fields);
  }
}

export function createResolvers(
  model: Model<any>,
  tc: ObjectTypeComposer<any, any>,
  opts: AllResolversOpts
): void {
  (Object.keys(allResolvers) as any).forEach((resolverName: keyof typeof allResolvers) => {
    if (!opts.hasOwnProperty(resolverName) || opts[resolverName] !== false) {
      const createResolverFn = allResolvers[resolverName] as any;
      if (typeof createResolverFn === 'function') {
        const resolver = createResolverFn(model, tc, opts[resolverName] || {});
        if (resolver) {
          tc.setResolver(resolverName, resolver);
        }
      }
    }
  });
}
