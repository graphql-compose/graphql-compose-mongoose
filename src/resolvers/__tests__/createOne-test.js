/* @flow */
/* eslint-disable no-param-reassign */

import { Resolver, TypeComposer, graphql } from 'graphql-compose';
import { UserModel } from '../../__mocks__/userModel';
import createOne from '../createOne';
import { composeWithMongoose } from '../../composeWithMongoose';
import GraphQLMongoID from '../../types/mongoid';
import typeStorage from '../../typeStorage';

const { GraphQLNonNull, GraphQLObjectType } = graphql;

beforeAll(() => UserModel.base.connect());
afterAll(() => UserModel.base.disconnect());

describe('createOne() ->', () => {
  let UserTypeComposer;

  beforeEach(() => {
    typeStorage.clear();
    UserModel.schema._gqcTypeComposer = undefined;
    UserTypeComposer = composeWithMongoose(UserModel);
  });

  beforeEach(async () => {
    await UserModel.remove({});
  });

  it('should return Resolver object', () => {
    const resolver = createOne(UserModel, UserTypeComposer);
    expect(resolver).toBeInstanceOf(Resolver);
  });

  describe('Resolver.args', () => {
    it('should have required `record` arg', () => {
      const resolver = createOne(UserModel, UserTypeComposer);
      const argConfig = resolver.getArg('record');
      expect(argConfig.type).toBeInstanceOf(GraphQLNonNull);
      // $FlowFixMe
      expect(argConfig.type.ofType.name).toBe('CreateOneUserInput');
    });
  });

  describe('Resolver.resolve():Promise', () => {
    it('should be promise', () => {
      const result = createOne(UserModel, UserTypeComposer).resolve({});
      expect(result).toBeInstanceOf(Promise);
      result.catch(() => 'catch error if appear, hide it from mocha');
    });

    it('should rejected with Error if args.record is empty', async () => {
      const result = createOne(UserModel, UserTypeComposer).resolve({ args: {} });
      await expect(result).rejects.toMatchSnapshot();
    });

    it('should return payload.recordId', async () => {
      const result = await createOne(UserModel, UserTypeComposer).resolve({
        args: {
          record: { name: 'newName' },
        },
      });
      expect(result.recordId).toBeTruthy();
    });

    it('should create document with args.record', async () => {
      const result = await createOne(UserModel, UserTypeComposer).resolve({
        args: {
          record: { name: 'newName' },
        },
      });
      expect(result.record.name).toBe('newName');
    });

    it('should save document to database', done => {
      const checkedName = 'nameForMongoDB';
      createOne(UserModel, UserTypeComposer)
        .resolve({
          args: {
            record: { name: checkedName },
          },
        })
        .then(res => {
          UserModel.collection.findOne({ _id: res.record._id }, (err, doc) => {
            expect(doc.name).toBe(checkedName);
            done();
          });
        });
    });

    it('should return payload.record', async () => {
      const result = await createOne(UserModel, UserTypeComposer).resolve({
        args: {
          record: { name: 'NewUser' },
        },
      });
      expect(result.record.id).toBe(result.recordId);
    });

    it('should return mongoose document', async () => {
      const result = await createOne(UserModel, UserTypeComposer).resolve({
        args: { record: { name: 'NewUser' } },
      });
      expect(result.record).toBeInstanceOf(UserModel);
    });

    it('should call `beforeRecordMutate` method with created `record` and `resolveParams` as args', async () => {
      const result = await createOne(UserModel, UserTypeComposer).resolve({
        args: { record: { name: 'NewUser' } },
        context: { ip: '1.1.1.1' },
        beforeRecordMutate: (record, rp) => {
          record.name = 'OverridedName';
          record.someDynamic = rp.context.ip;
          return record;
        },
      });
      expect(result.record).toBeInstanceOf(UserModel);
      expect(result.record.name).toBe('OverridedName');
      expect(result.record.someDynamic).toBe('1.1.1.1');
    });
  });

  describe('Resolver.getType()', () => {
    it('should have correct output type name', () => {
      const outputType = createOne(UserModel, UserTypeComposer).getType();
      // $FlowFixMe
      expect(outputType.name).toBe(`CreateOne${UserTypeComposer.getTypeName()}Payload`);
    });

    it('should have recordId field', () => {
      const outputType = createOne(UserModel, UserTypeComposer).getType();
      // $FlowFixMe
      const recordIdField = new TypeComposer(outputType).getField('recordId');
      expect(recordIdField.type).toBe(GraphQLMongoID);
    });

    it('should have record field', () => {
      const outputType = createOne(UserModel, UserTypeComposer).getType();
      // $FlowFixMe
      const recordField = new TypeComposer(outputType).getField('record');
      expect(recordField.type).toBe(UserTypeComposer.getType());
    });

    it('should reuse existed outputType', () => {
      const outputTypeName = `CreateOne${UserTypeComposer.getTypeName()}Payload`;
      const existedType = new GraphQLObjectType({
        name: outputTypeName,
        fields: () => ({}),
      });
      typeStorage.set(outputTypeName, existedType);
      const outputType = createOne(UserModel, UserTypeComposer).getType();
      expect(outputType).toBe(existedType);
    });
  });
});
