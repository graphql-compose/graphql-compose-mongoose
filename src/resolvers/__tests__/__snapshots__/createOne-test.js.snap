// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`createOne() -> Resolver.resolve():Promise should rejected with Error if args.record is empty 1`] = `[Error: User.createOne resolver requires at least one value in args.record]`;
